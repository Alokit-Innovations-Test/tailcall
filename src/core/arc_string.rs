use std::sync::Arc;

use headers::HeaderValue;
use schemars::gen::SchemaGenerator;
use schemars::schema::Schema;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ArcString(Arc<str>);

impl<'de> Deserialize<'de> for ArcString {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(ArcString::from)
    }
}

impl Serialize for ArcString {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.as_str().serialize(serializer)
    }
}

impl Default for ArcString {
    fn default() -> Self {
        ArcString("".into())
    }
}

impl AsRef<str> for ArcString {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<&str> for ArcString {
    fn from(s: &str) -> Self {
        ArcString(s.into())
    }
}

impl From<&String> for ArcString {
    fn from(s: &String) -> Self {
        ArcString(s.clone().into())
    }
}

impl From<String> for ArcString {
    fn from(s: String) -> Self {
        ArcString(s.into())
    }
}

impl From<ArcString> for Vec<u8> {
    fn from(s: ArcString) -> Self {
        s.as_str().as_bytes().to_vec()
    }
}

impl TryFrom<ArcString> for HeaderValue {
    type Error = <HeaderValue as TryFrom<String>>::Error;

    fn try_from(s: ArcString) -> Result<Self, Self::Error> {
        s.as_str().parse()
    }
}

impl PartialEq<&str> for ArcString {
    fn eq(&self, other: &&str) -> bool {
        self.as_ref() == *other
    }
}

impl schemars::JsonSchema for ArcString {
    fn schema_name() -> String {
        <String as schemars::JsonSchema>::schema_name()
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        <String as schemars::JsonSchema>::json_schema(gen)
    }
}

impl std::fmt::Display for ArcString {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        self.as_str().fmt(f)
    }
}

impl ArcString {
    pub fn as_str(&self) -> &str {
        self.as_ref()
    }
}
