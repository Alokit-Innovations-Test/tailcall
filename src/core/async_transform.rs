use std::future::Future;

use super::valid::{Valid, Validator};

/// A configuration transformer that allows us to perform various
/// transformations on the configuration before it's further processed for
/// blueprint creation.
pub trait AsyncTransform {
    type Value: Send;
    type Error;

    fn transform(
        &self,
        value: Self::Value,
    ) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send;
}

/// A suite of common operators that are available for all transformers.
pub trait AsyncTransformerOps: AsyncTransform + Send {
    fn pipe<Other: AsyncTransform<Value = Self::Value, Error = Self::Error>>(
        self,
        other: Other,
    ) -> Pipe<Self, Other>
    where
        Self: Sized,
    {
        Pipe(self, other)
    }

    fn generate(&self) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send
    where
        Self: Send + Sync,
        Self::Value: std::default::Default,
    {
        async move { self.transform(Self::Value::default()).await }
    }

    fn when(self, cond: bool) -> When<Self>
    where
        Self: Sized,
    {
        When(self, cond)
    }
}

impl<T: AsyncTransform + Sync + Send> AsyncTransformerOps for T {}

pub struct When<A>(A, bool);
impl<A: AsyncTransform + Sync + Send> AsyncTransform for When<A> {
    type Value = A::Value;
    type Error = A::Error;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        if self.1 {
            self.0.transform(value).await
        } else {
            Valid::succeed(value)
        }
    }
}

/// Represents a composition of two transformers.
pub struct Pipe<A, B>(A, B);

impl<A, E, X, Y> AsyncTransform for Pipe<X, Y>
where
    X: AsyncTransform<Value = A, Error = E> + Send + Sync,
    Y: AsyncTransform<Value = A, Error = E> + Send + Sync,
    A: Sync + Send,
    E: Sync + Send,
{
    type Value = A;
    type Error = E;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        let result = self.0.transform(value).await;
        match result.to_result() {
            Ok(result) => self.1.transform(result).await,
            Err(err) => Valid::from_validation_err(err),
        }
    }
}

/// Represents an empty transformer.
pub struct Default<A, E>(std::marker::PhantomData<(A, E)>);

impl<A: Send + Sync, E: Send + Sync> AsyncTransform for Default<A, E> {
    type Value = A;
    type Error = E;
    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> { Valid::succeed(value) }
}

// pub fn default<A, E>() -> Default<A, E> {
//     Default(std::marker::PhantomData)
// }
