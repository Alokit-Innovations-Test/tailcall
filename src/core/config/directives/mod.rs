use lazy_static::lazy_static;

mod add_field;
mod alias;
mod cache;
mod call;
mod expr;
mod federation;
mod graphql;
mod grpc;
mod http;
mod js;
mod link;
mod modify;
mod omit;
mod protected;
mod server;
mod telemetry;
mod upstream;

pub use add_field::*;
pub use alias::*;
pub use cache::*;
pub use call::*;
pub use expr::*;
pub use federation::*;
pub use graphql::*;
pub use grpc::*;
pub use http::*;
pub use js::*;
pub use link::*;
pub use modify::*;
pub use omit::*;
pub use protected::*;
pub use server::*;
pub use telemetry::*;
pub use upstream::*;

use crate::core::directive::DirectiveCodec;

lazy_static! {
    pub static ref KNOWN_DIRECTIVES: Vec<String> = vec![
        add_field::AddField::directive_name(),
        alias::Alias::directive_name(),
        cache::Cache::directive_name(),
        call::Call::directive_name(),
        expr::Expr::directive_name(),
        graphql::GraphQL::directive_name(),
        grpc::Grpc::directive_name(),
        http::Http::directive_name(),
        js::JS::directive_name(),
        link::Link::directive_name(),
        modify::Modify::directive_name(),
        omit::Omit::directive_name(),
        protected::Protected::directive_name(),
        server::Server::directive_name(),
        telemetry::Telemetry::directive_name(),
        upstream::Upstream::directive_name(),
    ];
}
