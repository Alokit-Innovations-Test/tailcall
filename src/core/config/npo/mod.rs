mod identifier;
mod queries;

use std::{collections::HashSet, rc::Rc};

pub use identifier::*;
pub use queries::*;

use super::Config;

struct NPOIdentifier<'a> {
    config: &'a Config,
}

impl<'a> NPOIdentifier<'a> {
    pub fn new(config: &'a Config) -> NPOIdentifier {
        NPOIdentifier { config }
    }

    fn iter(
        &self,
        path: Chunk<FieldName<'a>>,
        type_name: TypeName<'a>,
        is_list: bool,
        visited: &mut HashSet<TypeName<'a>>,
    ) -> Chunk<Chunk<FieldName<'a>>> {
        if visited.contains(&type_name) {
            return Chunk::new();
        } else {
            visited.insert(type_name);
        }
        let mut chunks = Chunk::new();
        if let Some(type_of) = self.config.find_type(&type_name.as_str()) {
            for (name, field) in type_of.fields.iter() {
                let path = path.clone().append(FieldName::new(name));
                let is_batch = field.has_batched_resolver();

                if field.has_resolver() {
                    if !is_batch && is_list {
                        chunks = chunks.append(path.clone());
                    }
                }

                let is_list = is_list | field.list;

                chunks = chunks.concat(self.iter(
                    path,
                    TypeName::new(field.type_of.as_str()),
                    is_list,
                    visited,
                ))
            }
        }

        chunks
    }

    fn find_chunks(&self) -> Chunk<Chunk<FieldName<'a>>> {
        match &self.config.schema.query {
            None => Chunk::new(),
            Some(query) => self.iter(
                Chunk::new(),
                TypeName::new(query.as_str()),
                false,
                &mut HashSet::new(),
            ),
        }
    }

    fn find(&self) -> Vec<Vec<&'a str>> {
        self.find_chunks()
            .as_vec()
            .iter()
            .map(|chunk| {
                chunk
                    .as_vec()
                    .iter()
                    .map(|field_name| field_name.as_str())
                    .collect()
            })
            .collect()
    }
}

#[derive(Clone)]
enum Chunk<A> {
    Nil,
    Cons(A, Rc<Chunk<A>>),
    Concat(Rc<Chunk<A>>, Rc<Chunk<A>>),
}

impl<A> Chunk<A> {
    fn new() -> Self {
        Self::Nil
    }

    fn from_vec(vec: Vec<A>) -> Chunk<A> {
        let mut path = Chunk::new();
        for a in vec.into_iter().rev() {
            path = Chunk::Cons(a, Rc::new(path));
        }
        path
    }

    fn append(self, a: A) -> Self {
        Chunk::Cons(a, Rc::new(self))
    }

    fn concat(self, other: Chunk<A>) -> Self {
        Self::Concat(Rc::new(self), Rc::new(other))
    }

    fn as_vec(&self) -> Vec<&A> {
        let mut vec = Vec::new();
        self.as_vec_mut(&mut vec);
        vec
    }

    fn as_vec_mut<'a>(&'a self, buf: &mut Vec<&'a A>) {
        match self {
            Chunk::Nil => {}
            Chunk::Cons(a, rest) => {
                buf.push(a);
                rest.as_vec_mut(buf);
            }
            Chunk::Concat(a, b) => {
                a.as_vec_mut(buf);
                b.as_vec_mut(buf);
            }
        }
    }
}
