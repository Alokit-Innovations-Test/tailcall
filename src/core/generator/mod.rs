mod from_json;
pub mod config;
mod from_proto;
mod generator;
mod graphql_type;
mod json;
mod proto;
mod from_generalized_config;
pub mod source;

pub use from_json::{from_json, ConfigGenerationRequest};
pub use json::NameGenerator;
pub use generator::Generator;
pub use from_generalized_config::FromGeneralizedConfig;
