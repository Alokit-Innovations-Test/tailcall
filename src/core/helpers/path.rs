use std::fs;
use std::path::Path;

use pathdiff::diff_paths;

/// Checks if file or folder already exists or not.
pub fn is_exists(path: &str) -> bool {
    fs::metadata(path).is_ok()
}

/// Expects both paths to be absolute and returns a relative path from `from` to
/// `to`. expects `from`` to be directory.
pub fn to_relative_path(from: &Path, to: &str) -> Option<String> {
    let from_path = Path::new(from).to_path_buf();
    let to_path = Path::new(to).to_path_buf();

    // Calculate the relative path from `from_path` to `to_path`
    diff_paths(to_path, from_path).map(|p| p.to_string_lossy().to_string())
}
