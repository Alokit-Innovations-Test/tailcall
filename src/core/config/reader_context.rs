use std::borrow::Cow;
use std::collections::BTreeMap;

use headers::HeaderMap;
use serde::Deserialize;
use crate::core::has_headers::HasHeaders;
use crate::core::json::JsonLike;
use crate::core::path::PathString;
use crate::core::runtime::TargetRuntime;

pub struct ConfigReaderContext<'a, Value> {
    pub runtime: &'a TargetRuntime<Value>,
    pub vars: &'a BTreeMap<String, String>,
    pub headers: HeaderMap,
}

impl<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone> PathString for ConfigReaderContext<'a, Value> {
    fn path_string<T: AsRef<str>>(&self, path: &[T]) -> Option<Cow<'_, str>> {
        if path.is_empty() {
            return None;
        }

        path.split_first()
            .and_then(|(head, tail)| match head.as_ref() {
                "vars" => self.vars.get(tail[0].as_ref()).map(|v| v.into()),
                "env" => self.runtime.env.get(tail[0].as_ref()),
                _ => None,
            })
    }
}

impl<Value> HasHeaders for ConfigReaderContext<'_, Value> {
    fn headers(&self) -> &HeaderMap {
        &self.headers
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use crate::core::tests::TestEnvIO;

    #[test]
    fn path_string() {
        let mut runtime = crate::core::runtime::test::init(None);
        runtime.env = Arc::new(TestEnvIO::from_iter([(
            "ENV_1".to_owned(),
            "ENV_VAL".to_owned(),
        )]));

        let reader_context = ConfigReaderContext {
            runtime: &runtime,
            vars: &BTreeMap::from_iter([("VAR_1".to_owned(), "VAR_VAL".to_owned())]),
            headers: Default::default(),
        };

        assert_eq!(
            reader_context.path_string(&["env", "ENV_1"]),
            Some("ENV_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["env", "ENV_5"]), None);
        assert_eq!(
            reader_context.path_string(&["vars", "VAR_1"]),
            Some("VAR_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["vars", "VAR_6"]), None);
        assert_eq!(reader_context.path_string(&["unknown", "unknown"]), None);
    }
}
