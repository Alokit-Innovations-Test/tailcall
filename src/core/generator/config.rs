use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
