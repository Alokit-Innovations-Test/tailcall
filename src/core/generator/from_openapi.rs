use std::collections::BTreeMap;

use oas3::Spec;

use crate::core::config::{Config, Enum, RootSchema, Type, Union, Upstream};

#[derive(Default)]
pub struct OpenApiToConfigConverter {
    pub query: String,
    pub spec: Spec,
    pub types: BTreeMap<String, Type>,
    pub unions: BTreeMap<String, Union>,
    pub enums: BTreeMap<String, Enum>,
}

impl OpenApiToConfigConverter {
    pub fn new(query: impl AsRef<str>, spec_str: impl AsRef<str>) -> anyhow::Result<Self> {
        let spec = oas3::from_reader(spec_str.as_ref().as_bytes())?;
        Ok(Self {
            query: query.as_ref().to_string(),
            spec,
            ..Default::default()
        })
    }

    pub fn convert(self) -> Config {
        let config = Config {
            server: Default::default(),
            upstream: Upstream {
                base_url: self.spec.servers.first().cloned().map(|server| server.url),
                ..Default::default()
            },
            schema: RootSchema {
                query: self.types.get(&self.query).map(|_| self.query),
                ..Default::default()
            },
            types: self.types,
            unions: self.unions,
            enums: self.enums,
            ..Default::default()
        };
        config
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use super::*;

    #[test]
    fn test_openapi_apis_guru() {
        let apis_guru = config_from_openapi_spec("apis-guru.yml");
        insta::assert_snapshot!(apis_guru);
    }

    #[test]
    fn test_openapi_jsonplaceholder() {
        let jsonplaceholder = config_from_openapi_spec("jsonplaceholder.yml");
        insta::assert_snapshot!(jsonplaceholder);
    }

    #[test]
    fn test_openapi_spotify() {
        let spotify = config_from_openapi_spec("spotify.yml");
        insta::assert_snapshot!(spotify);
    }

    fn config_from_openapi_spec(filename: &str) -> String {
        let spec_path = Path::new("src")
            .join("core")
            .join("generator")
            .join("openapi")
            .join(filename);

        let content = fs::read_to_string(spec_path).unwrap();
        OpenApiToConfigConverter::new("Query", content.as_str())
            .unwrap()
            .convert()
            .to_sdl()
    }
}
