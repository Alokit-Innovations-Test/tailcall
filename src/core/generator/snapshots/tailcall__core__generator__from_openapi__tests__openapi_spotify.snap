---
source: src/core/generator/from_openapi.rs
expression: spotify
---
schema @server @upstream(baseURL: "https://api.spotify.com/v1") {
  query: Query
}

type Query {
  """
  Check to see if the current user is following one or more artists or other Spotify users.

  """
  checkCurrentUserFollows(ids: String!, type: Type13!): [Boolean] @http(path: "/me/following/contains")
  """
  Check to see if one or more Spotify users are following a specified playlist.

  """
  checkIfUserFollowsPlaylist(ids: String!, playlist_id: String!): [Boolean] @http(path: "/playlists/{playlist_id}/followers/contains")
  """
  Check if one or more albums is already saved in the current Spotify user's 'Your Music' library.

  """
  checkUsersSavedAlbums(ids: String!): [Boolean] @http(path: "/me/albums/contains")
  """
  Check if one or more audiobooks are already saved in the current Spotify user's library.

  """
  checkUsersSavedAudiobooks(ids: String!): [Boolean] @http(path: "/me/audiobooks/contains")
  """
  Check if one or more episodes is already saved in the current Spotify user's 'Your Episodes' library.<br/>
This API endpoint is in __beta__ and could change without warning. Please share any feedback that you have, or issues that you discover, in our [developer community forum](https://community.spotify.com/t5/Spotify-for-Developers/bd-p/Spotify_Developer)..

  """
  checkUsersSavedEpisodes(ids: String!): [Boolean] @http(path: "/me/episodes/contains")
  """
  Check if one or more shows is already saved in the current Spotify user's library.

  """
  checkUsersSavedShows(ids: String!): [Boolean] @http(path: "/me/shows/contains")
  """
  Check if one or more tracks is already saved in the current Spotify user's 'Your Music' library.

  """
  checkUsersSavedTracks(ids: String!): [Boolean] @http(path: "/me/tracks/contains")
  """
  Get a list of Spotify playlists tagged with a particular category.

  """
  getACategoriesPlaylists(category_id: String!, limit: Int, offset: Int): PagingFeaturedPlaylistObject @http(path: "/browse/categories/{category_id}/playlists")
  """
  Get a single category used to tag items in Spotify (on, for example, the Spotify player’s “Browse” tab).

  """
  getACategory(category_id: String!, locale: String): CategoryObject @http(path: "/browse/categories/{category_id}")
  """
  Get Spotify catalog information for a single audiobook chapter. Chapters are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.

  """
  getAChapter(id: String!, market: String): ChapterObject @http(path: "/chapters/{id}")
  """
  Get a list of the playlists owned or followed by the current Spotify
user.

  """
  getAListOfCurrentUsersPlaylists(limit: Int, offset: Int): PagingPlaylistObject @http(path: "/me/playlists")
  """
  Get Spotify catalog information for a single show identified by its
unique Spotify ID.

  """
  getAShow(id: String!, market: String): ShowObject @http(path: "/shows/{id}")
  """
  Get Spotify catalog information about an show’s episodes. Optional parameters can be used to limit the number of episodes returned.

  """
  getAShowsEpisodes(id: String!, limit: Int, market: String, offset: Int): PagingSimplifiedEpisodeObject @http(path: "/shows/{id}/episodes")
  """
  Get information about a user’s available Spotify Connect devices. Some device models are not supported and will not be listed in the API response.

  """
  getAUsersAvailableDevices: Type14 @http(path: "/me/player/devices")
  """
  Get Spotify catalog information for a single album.

  """
  getAnAlbum(id: String!, market: String): AlbumObject @http(path: "/albums/{id}")
  """
  Get Spotify catalog information about an album’s tracks.
Optional parameters can be used to limit the number of tracks returned.

  """
  getAnAlbumsTracks(id: String!, limit: Int, market: String, offset: Int): PagingSimplifiedTrackObject @http(path: "/albums/{id}/tracks")
  """
  Get Spotify catalog information for a single artist identified by their unique Spotify ID.

  """
  getAnArtist(id: String!): ArtistObject @http(path: "/artists/{id}")
  """
  Get Spotify catalog information about an artist's albums.

  """
  getAnArtistsAlbums(id: String!, include_groups: String, limit: Int, market: String, offset: Int): PagingArtistDiscographyAlbumObject @http(path: "/artists/{id}/albums")
  """
  Get Spotify catalog information about artists similar to a given artist. Similarity is based on analysis of the Spotify community's listening history.

  """
  getAnArtistsRelatedArtists(id: String!): Type2 @http(path: "/artists/{id}/related-artists")
  """
  Get Spotify catalog information about an artist's top tracks by country.

  """
  getAnArtistsTopTracks(id: String!, market: String): Type3 @http(path: "/artists/{id}/top-tracks")
  """
  Get Spotify catalog information for a single audiobook. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.

  """
  getAnAudiobook(id: String!, market: String): AudiobookObject @http(path: "/audiobooks/{id}")
  """
  Get Spotify catalog information for a single episode identified by its
unique Spotify ID.

  """
  getAnEpisode(id: String!, market: String): EpisodeObject @http(path: "/episodes/{id}")
  """
  Get a low-level audio analysis for a track in the Spotify catalog. The audio analysis describes the track’s structure and musical content, including rhythm, pitch, and timbre.

  """
  getAudioAnalysis(id: String!): AudioAnalysisObject @http(path: "/audio-analysis/{id}")
  """
  Get audio feature information for a single track identified by its unique
Spotify ID.

  """
  getAudioFeatures(id: String!): AudioFeaturesObject @http(path: "/audio-features/{id}")
  """
  Get Spotify catalog information about an audiobook's chapters. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.

  """
  getAudiobookChapters(id: String!, limit: Int, market: String, offset: Int): PagingSimplifiedChapterObject @http(path: "/audiobooks/{id}/chapters")
  """
  Get the list of markets where Spotify is available.

  """
  getAvailableMarkets: Type10 @http(path: "/markets")
  """
  Get a list of categories used to tag items in Spotify (on, for example, the Spotify player’s “Browse” tab).

  """
  getCategories(limit: Int, locale: String, offset: Int): Type6 @http(path: "/browse/categories")
  """
  Get detailed profile information about the current user (including the
current user's username).

  """
  getCurrentUsersProfile: PrivateUserObject @http(path: "/me")
  """
  Get a list of Spotify featured playlists (shown, for example, on a Spotify player's 'Browse' tab).

  """
  getFeaturedPlaylists(limit: Int, locale: String, offset: Int): PagingFeaturedPlaylistObject @http(path: "/browse/featured-playlists")
  """
  Get the current user's followed artists.

  """
  getFollowed(after: String, limit: Int, type: Type11!): Type12 @http(path: "/me/following")
  """
  Get information about the user’s current playback state, including track or episode, progress, and active device.

  """
  getInformationAboutTheUsersCurrentPlayback(additional_types: String, market: String): CurrentlyPlayingContextObject @http(path: "/me/player")
  """
  Get a list of the playlists owned or followed by a Spotify user.

  """
  getListUsersPlaylists(limit: Int, offset: Int, user_id: String!): PagingPlaylistObject @http(path: "/users/{user_id}/playlists")
  """
  Get Spotify catalog information for multiple albums identified by their Spotify IDs.

  """
  getMultipleAlbums(ids: String!, market: String): Type0 @http(path: "/albums")
  """
  Get Spotify catalog information for several artists based on their Spotify IDs.

  """
  getMultipleArtists(ids: String!): Type1 @http(path: "/artists")
  """
  Get Spotify catalog information for several audiobooks identified by their Spotify IDs. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.

  """
  getMultipleAudiobooks(ids: String!, market: String): Type5 @http(path: "/audiobooks")
  """
  Get Spotify catalog information for several episodes based on their Spotify IDs.

  """
  getMultipleEpisodes(ids: String!, market: String): Type9 @http(path: "/episodes")
  """
  Get Spotify catalog information for several shows based on their Spotify IDs.

  """
  getMultipleShows(ids: String!, market: String): Type19 @http(path: "/shows")
  """
  Get a list of new album releases featured in Spotify (shown, for example, on a Spotify player’s “Browse” tab).

  """
  getNewReleases(limit: Int, offset: Int): Type7 @http(path: "/browse/new-releases")
  """
  Get a playlist owned by a Spotify user.

  """
  getPlaylist(additional_types: String, fields: String, market: String, playlist_id: String!): PlaylistObject @http(path: "/playlists/{playlist_id}")
  """
  Get the current image associated with a specific playlist.

  """
  getPlaylistCover(playlist_id: String!): [ImageObject] @http(path: "/playlists/{playlist_id}/images")
  """
  Get full details of the items of a playlist owned by a Spotify user.

  """
  getPlaylistsTracks(additional_types: String, fields: String, limit: Int, market: String, offset: Int, playlist_id: String!): PagingPlaylistTrackObject @http(path: "/playlists/{playlist_id}/tracks")
  """
  Get the list of objects that make up the user's queue.

  """
  getQueue: QueueObject @http(path: "/me/player/queue")
  """
  Get tracks from the current user's recently played tracks.
_**Note**: Currently doesn't support podcast episodes._

  """
  getRecentlyPlayed(after: Int, before: Int, limit: Int): CursorPagingPlayHistoryObject @http(path: "/me/player/recently-played")
  """
  Retrieve a list of available genres seed parameter values for [recommendations](/documentation/web-api/reference/get-recommendations).

  """
  getRecommendationGenres: Type15 @http(path: "/recommendations/available-genre-seeds")
  """
  Recommendations are generated based on the available information for a given seed entity and matched against similar artists and tracks. If there is sufficient information about the provided seeds, a list of tracks will be returned together with pool size details.

For artists and tracks that are very new or obscure there might not be enough data to generate a list of tracks.

  """
  getRecommendations(limit: Int, market: String, max_acousticness: Int, max_danceability: Int, max_duration_ms: Int, max_energy: Int, max_instrumentalness: Int, max_key: Int, max_liveness: Int, max_loudness: Int, max_mode: Int, max_popularity: Int, max_speechiness: Int, max_tempo: Int, max_time_signature: Int, max_valence: Int, min_acousticness: Int, min_danceability: Int, min_duration_ms: Int, min_energy: Int, min_instrumentalness: Int, min_key: Int, min_liveness: Int, min_loudness: Int, min_mode: Int, min_popularity: Int, min_speechiness: Int, min_tempo: Int, min_time_signature: Int, min_valence: Int, seed_artists: String, seed_genres: String, seed_tracks: String, target_acousticness: Int, target_danceability: Int, target_duration_ms: Int, target_energy: Int, target_instrumentalness: Int, target_key: Int, target_liveness: Int, target_loudness: Int, target_mode: Int, target_popularity: Int, target_speechiness: Int, target_tempo: Int, target_time_signature: Int, target_valence: Int): RecommendationsObject @http(path: "/recommendations")
  """
  Get audio features for multiple tracks based on their Spotify IDs.

  """
  getSeveralAudioFeatures(ids: String!): Type4 @http(path: "/audio-features")
  """
  Get Spotify catalog information for several audiobook chapters identified by their Spotify IDs. Chapters are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.

  """
  getSeveralChapters(ids: String!, market: String): Type8 @http(path: "/chapters")
  """
  Get Spotify catalog information for multiple tracks based on their Spotify IDs.

  """
  getSeveralTracks(ids: String!, market: String): Type20 @http(path: "/tracks")
  """
  Get the object currently being played on the user's Spotify account.

  """
  getTheUsersCurrentlyPlayingTrack(additional_types: String, market: String): CurrentlyPlayingObject @http(path: "/me/player/currently-playing")
  """
  Get Spotify catalog information for a single track identified by its
unique Spotify ID.

  """
  getTrack(id: String!, market: String): TrackObject @http(path: "/tracks/{id}")
  """
  Get public profile information about a Spotify user.

  """
  getUsersProfile(user_id: String!): PublicUserObject @http(path: "/users/{user_id}")
  """
  Get a list of the albums saved in the current Spotify user's 'Your Music' library.

  """
  getUsersSavedAlbums(limit: Int, market: String, offset: Int): PagingSavedAlbumObject @http(path: "/me/albums")
  """
  Get a list of the audiobooks saved in the current Spotify user's 'Your Music' library.

  """
  getUsersSavedAudiobooks(limit: Int, offset: Int): PagingSavedAudiobookObject @http(path: "/me/audiobooks")
  """
  Get a list of the episodes saved in the current Spotify user's library.<br/>
This API endpoint is in __beta__ and could change without warning. Please share any feedback that you have, or issues that you discover, in our [developer community forum](https://community.spotify.com/t5/Spotify-for-Developers/bd-p/Spotify_Developer).

  """
  getUsersSavedEpisodes(limit: Int, market: String, offset: Int): PagingSavedEpisodeObject @http(path: "/me/episodes")
  """
  Get a list of shows saved in the current Spotify user's library. Optional parameters can be used to limit the number of shows returned.

  """
  getUsersSavedShows(limit: Int, offset: Int): PagingSavedShowObject @http(path: "/me/shows")
  """
  Get a list of the songs saved in the current Spotify user's 'Your Music' library.

  """
  getUsersSavedTracks(limit: Int, market: String, offset: Int): PagingSavedTrackObject @http(path: "/me/tracks")
  """
  Get the current user's top artists based on calculated affinity.

  """
  getUsersTopArtists(limit: Int, offset: Int, time_range: String): PagingArtistObject @http(path: "/me/top/artists")
  """
  Get the current user's top tracks based on calculated affinity.

  """
  getUsersTopTracks(limit: Int, offset: Int, time_range: String): PagingTrackObject @http(path: "/me/top/tracks")
  """
  Get Spotify catalog information about albums, artists, playlists, tracks, shows, episodes or audiobooks
that match a keyword string. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.

  """
  search(include_external: Type17, limit: Int, market: String, offset: Int, q: String!, type: [Type16]!): Type18 @http(path: "/search")
}
